data/src/main/scala/cats/data/Or.scala

```diff
@@ -1,506 +1,124 @@
-  def @\?/[AA, BB](k: Validation[A, B] => Validation[AA, BB]): AA Or BB =
-    validationed(k)
+  def traverse[F[_], AA >: A, D](f: B => F[D])(implicit F: Applicative[F]): F[AA Or D] = this match {
+    case RightOr(b) => F.map(f(b))(Or.right _)
+    case left @ LeftOr(_) => F.pure(left)
+  }
 
-  /** Return the value from whichever side of the disjunction is defined, given a commonly assignable type. */
-  def merge[AA >: A](implicit ev: B <~< AA): AA =
-    this match {
-      case LOr(a) => a
-      case ROr(b) => ev(b)
-    }
+  def foldLeft[C](c: C)(f: (C, B) => C): C = fold(_ => c, f(c, _))
 
-  /** Convert to a These. */
-  def toThese: A \&/ B =
-    fold(
-      a => \&/.This(a),
-      b => \&/.That(b)
-    )
-    */
+  def foldRight[C](c: C)(f: (B, C) => C): C = fold(_ => c, f(_, c))
 
+  def foldRight[C](c: Lazy[C])(f: (B, Lazy[C]) => C): Lazy[C] = fold(_ => c, b => Lazy(f(b, c)))
 }
 
-/** A left disjunction
- *
- * Often used to represent the failure case of a result
- */
-final case class LOr[+A](a: A) extends (A Or Nothing)
+object Or extends OrFunctions {
+  case class LeftOr[+A](a: A) extends (A Or Nothing)
 
-/** A right disjunction
- *
- * Often used to represent the success case of a result
- */
-final case class ROr[+B](b: B) extends (Nothing Or B)
-
-object Or extends OrInstances with OrFunctions {
-
-  /** Spin in tail-position on the right value of the given disjunction. */
-  @annotation.tailrec
-  final def loopRight[A, B, X](d: A Or B, left: A => X, right: B => X Or (A Or B)): X =
-    d match {
-      case LOr(a) => left(a)
-      case ROr(b) => right(b) match {
-        case LOr(x) => x
-        case ROr(q) => loopRight(q, left, right)
-      }
-    }
+  case class RightOr[+B](b: B) extends (Nothing Or B)
 
-  /** Spin in tail-position on the left value of the given disjunction. */
-  @annotation.tailrec
-  final def loopLeft[A, B, X](d: A Or B, left: A => X Or (A Or B), right: B => X): X =
-    d match {
-      case LOr(a) => left(a) match {
-        case LOr(x) => x
-        case ROr(q) => loopLeft(q, left, right)
-      }
-      case ROr(b) => right(b)
-    }
-
-}
-
-sealed abstract class OrInstances extends OrInstances0 {
-  implicit def OrOrder[A: Order, B: Order]: Order[A Or B] =
-    new Order[A Or B] {
-      def compare(a1: A Or B, a2: A Or B) =
-        a1 compare a2
-      override def eqv(a1: A Or B, a2: A Or B) =
-        a1 === a2
-    }
-
-  implicit def OrMonoid[A: Semigroup, B: Monoid]: Monoid[A Or B] =
-    new Monoid[A Or B] {
-      def combine(a1: A Or B, a2: A Or B) =
-        a1 +++ a2
-      def empty =
-        ROr(Monoid[B].empty)
-    }
-}
-
-sealed abstract class OrInstances0 extends OrInstances1 {
-  implicit def OrEqual[A: Eq, B: Eq]: Eq[A Or B] =
-    new Eq[A Or B] {
-      def eqv(a1: A Or B, a2: A Or B) =
-        a1 === a2
+  implicit class MergeableOr[A](private val self: A Or A) extends AnyVal {
```

Comment:
This is interesting. Why is this is done with syntax rather than a method with a type constraint? i.e., something like

```scala
def merge[X >: A](implicit ev: B <:< X]): X = ...
```

Apologies if I missed an earlier discussion about this.